### PR TITLE
Create services for pdf to text and training data generation

### PR DESCRIPTION
**Pull request by jenlu33**
[Asana ticket](https://app.asana.com/0/1203289004376659/1208127435414321/f)

I created a new service that uses PDF Reader to convert pdf to text and then is used in `GenerateTrainingData`. Right now it just returns work process information, but will probably need to adjust the prompt to get the rest of the data for an occupation standard.

Here is an example response:
```
[
  {
    "title": "Proper/safe use of tools, materials, and equipment",
    "description": "Basic awareness as well as use of industry tools & equipment, Common materials (and alternatives) used in the industry, Proper storage & disposal of common industry materials, Proper maintenance of industry tools & equipment",
    "defaultHours": null,
    "minimumHours": 1060,
    "maximumHours": 1330,
    "competencies": [
      {
        "title": "Basic awareness as well as use of industry tools & equipment"
      },
      {
        "title": "Common materials (and alternatives) used in the industry"
      },
      {
        "title": "Proper storage & disposal of common industry materials"
      },
      {
        "title": "Proper maintenance of industry tools & equipment"
      }
    ]
  },
  {
    "title": "Hazard awareness, safe work practices, and personal protective equipment",
    "description": "Introduction to construction safety (OSHA 10 or equivalent), Identification & use of common industry personal protective equipment, Hazardous material handling and storage (including SDS utilization), Situational awareness for the protection of self & others",
    "defaultHours": null,
    "minimumHours": 1063,
    "maximumHours": 1334,
    "competencies": [
      {
        "title": "Introduction to construction safety (OSHA 10 or equivalent)"
      },
      {
        "title": "Identification & use of common industry personal protective equipment"
      },
      {
        "title": "Hazardous material handling and storage (including SDS utilization)"
      },
      {
        "title": "Situational awareness for the protection of self & others"
      }
    ]
  },
  {
    "title": "Job layouts, material planning, and job-related math",
    "description": "Basic construction & industry specific math, Blueprint & other industry drawing interpretation, Basic material planning & scheduling, Project management, scheduling, estimating and reporting",
    "defaultHours": null,
    "minimumHours": 1063,
    "maximumHours": 1334,
    "competencies": [
      {
        "title": "Basic construction & industry specific math"
      },
      {
        "title": "Blueprint & other industry drawing interpretation"
      },
      {
        "title": "Basic material planning & scheduling"
      },
      {
        "title": "Project management, scheduling, estimating and reporting"
      }
    ]
  },
  {
    "title": "Interaction with management, supervisors, and/or customers",
    "description": "Receiving company and/or project level guidance, Receiving task level directions, clarifying, and communicating needs, Responding to customer service inquiries or requests, Engaging in professional communication (speaking, written, electronic)",
    "defaultHours": null,
    "minimumHours": 1063,
    "maximumHours": 1334,
    "competencies": [
      {
        "title": "Receiving company and/or project level guidance"
      },
      {
        "title": "Receiving task level directions, clarifying, and communicating needs"
      },
      {
        "title": "Responding to customer service inquiries or requests"
      },
      {
        "title": "Engaging in professional communication (speaking, written, electronic)"
      }
    ]
  },
  {
    "title": "New installations/projects",
    "description": "Determining project requirements (customer, company, contractor), System planning, sizing, and configuration, Regulatory compliance (licenses, permits, inspections), Final documentation & hand-off to owner",
    "defaultHours": null,
    "minimumHours": 1063,
    "maximumHours": 1334,
    "competencies": [
      {
        "title": "Determining project requirements (customer, company, contractor)"
      },
      {
        "title": "System planning, sizing, and configuration"
      },
      {
        "title": "Regulatory compliance (licenses, permits, inspections)"
      },
      {
        "title": "Final documentation & hand-off to owner"
      }
    ]
  },
  {
    "title": "Troubleshooting, warranty, service and/or repair",
    "description": "Identification & compliance with manufacturer’s instructions, Warranty claim identification & preparation, Troubleshooting & diagnosis of system/project challenges, Final start-up & punch-out procedures",
    "defaultHours": null,
    "minimumHours": 1063,
    "maximumHours": 1334,
    "competencies": [
      {
        "title": "Identification & compliance with manufacturer’s instructions"
      },
      {
        "title": "Warranty claim identification & preparation"
      },
      {
        "title": "Troubleshooting & diagnosis of system/project challenges"
      },
      {
        "title": "Final start-up & punch-out procedures"
      }
    ]
  }
]
```

### CODE REVIEW

File: app/services/generate_training_data.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+class GenerateTrainingData
+  def initialize(import_id)
+    @client = OpenAI::Client.new
+    @import_id = import_id
+  end
+
+  def call
+    response = @client.chat(
+      parameters: {
+        model: "gpt-4",
```
Comment (jhash) [inline app/services/generate_training_data.rb:12]: Do we need to use gpt-4 or would 3.5 suffice?

Comment (jenlu33) [reply app/services/generate_training_data.rb:12]: 3.5 could probably be okay. I didn't test it out yet

File: app/services/generate_training_data.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  def extract_text
+    text = PdfToText.new(@import_id).call
+    text.gsub('\n', " ").gsub(/\s+/, " ").strip
+  end
+
+  def base_prompt
+    "Get the work processes from the following text in JSON format. JSON output needs the following fields:\
+      title: title of the work process.
+      description: description for the work process.
+      defaultHours: amount of hours. It is optional. If value not found, return null.
```
Comment (jhash) [inline app/services/generate_training_data.rb:34]: Can we add more examples where this value should be filled in and adjust the prompt if needed to return the correct value?

File: spec/services/generate_training_data_spec.rb [hunk 3]
```diff
@@ -0,0 +1,42 @@
+require "rails_helper"
+
+RSpec.describe GenerateTrainingData do
+  describe "#call" do
+    it "returns training data based on the import" do
```
Comment (jhash) [inline spec/services/generate_training_data_spec.rb:5]: Should there be expects somewhere in these tests?

File: app/services/generate_training_data.rb (hunk 2 above)
Comment (jenlu33) [reply app/services/generate_training_data.rb:34]: @jhash I'm not sure I know what you mean by this. `defaultHours` is just a number

**Review by FerPerales — APPROVED** [review]

**Pull request merged**